dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/writer/GeneWriter.java

```diff
@@ -17,139 +17,478 @@
+    if (startExon.isMissingNode() && endExon.isMissingNode()) {
+      transcript.putNull("start_exon");
+      transcript.putNull("end_exon");
+      return transcript;
+    } else if (startExon.isMissingNode()) {
+      transcript.putNull("start_exon");
+      return transcript;
+    } else if (endExon.isMissingNode()) {
+      transcript.putNull("end_exon");
+      return transcript;
+    }
+
+    double cdsLength = 0.0;
+    for (int i = startExon.asInt(); i <= endExon.asInt(); i++) {
+      val exon = (ObjectNode) exons.get(i);
+
+      // Initiate the values by assuming the whole exon is coding first
+      exon.put("genomic_coding_start", exon.get("start").asInt());
+      exon.put("cdna_coding_start", exon.get("cdna_start").asInt());
+      exon.put("genomic_coding_end", exon.get("end").asInt());
+      exon.put("cdna_coding_end", exon.get("cdna_end").asInt());
+
+      val cds = exon.path("cds");
+      if (!cds.isMissingNode()) {
+
+        // Translation id is the protein id of the coding sequence.
+        if (transcript.get("translation_id").isNull()) {
+          transcript.put("translation_id", cds.get("protein_id").asText());
+        }
+
+        cdsLength += cds.get("locationEnd").asInt() - cds.get("locationStart").asInt() + 1;
+      }
 
-        if (++insertCount % STATUS_GENE_COUNT == 0) {
-          log.info("Saved {} gene documents ({} documents/s)",
-              formatCount(insertCount), formatCount(STATUS_GENE_COUNT / (watch.elapsed(SECONDS))));
-          watch.reset().start();
+      // Start Exon.
+      if (i == startExon.asInt()) {
+        val start = exon.get("cds").get("locationStart").asInt();
+        if (strand.equals("-1")) {
+          val end = exon.get("cds").get("locationEnd").asInt();
+          transcript.put("coding_region_end", end);
+          exon.put("genomic_coding_end", end);
+          exon.put("cdna_coding_start", exon.get("end").asInt() - end + exon.get("cdna_start").asInt());
+        } else {
+          transcript.put("coding_region_start", start);
+          exon.put("genomic_coding_start", start);
+          exon.put("cdna_coding_start", exon.get("cdna_start").asInt() + (start - exon.get("start").asInt() + 1));
         }
+        transcript.put("cdna_coding_start", exon.get("cdna_coding_start").asInt());
+      }
+
+      // End Exon. Note: This can be the same exon as the start exon.
+      if (i == endExon.asInt()) {
+
+        if (strand.equals("-1")) {
+          if (cds.isMissingNode()) {
+            val start = exons.get(i - 1).path("cds").get("locationStart").asInt();
+            transcript.put("coding_region_start", start);
+          } else {
+            val start = cds.get("locationStart").asInt() + 1;
+            transcript.put("coding_region_start", start);
+            exon.put("genomic_coding_start", start);
+            exon.put("cdna_coding_end", exon.get("cdna_coding_start").asInt() + exon.get("genomic_coding_end").asInt()
```

Comment:
Some comments for things like this would be great. Helper methods will help to break it up and make it easier to read as well. ex. `normalizeXxxx`, `calculateXxxLength`, etc.